### PR TITLE
Fix buffer overflow in brainfuck assembler

### LINKED ISSUE #17746: Buffer overflow in brainfuck assembler
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Ubuntu 20.04.1 LTS x86_64
| File format of the file you reverse (mandatory)      | Brainfuck
| Architecture/bits of the file (mandatory)            | Brainfuck
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.6.0-git 25617 @ linux-x86-64 git.4.4.0-784-gb4a6b3ab2 commit: b4a6b3ab20ad20aaad6ff8b80fa6a9a122c8311f build: 2020-10-03__15:24:17

Example file contents: `++++`

### Expected behavior

```
[0x00000000]> oo+
[0x00000000]> wa out [ptr], 100
Written 100 byte(s) (out [ptr], 100) = wx 2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e2e
```

### Actual behavior

```
[0x00000000]> oo+
[0x00000000]> wa out [ptr], 100
*** stack smashing detected ***: terminated
```

### Additional Logs, screenshots, source-code,  configuration dump, ...
Probably related to line 68, the instruction needs to be just a little bit bigger than 32 bytes for the crash to occur. I have prepared a fix that dynamically allocates the buffer and I will open a pull request soon.
https://github.com/radareorg/radare2/blob/b4a6b3ab20ad20aaad6ff8b80fa6a9a122c8311f/libr/asm/p/asm_bf.c#L61-L70

### PR DESCRIPTION
**Pull request by plaets**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

The assembler was using a buffer stored on the stack to store data of the assembled instruction. Because of that, any instruction bigger than the size of that buffer (32 bytes) would corrupt the stack.

**Test plan**

1. Open any brainfuck file in write mode
2. Use `wa out [ptr], 1000` or any other command that will assemble and write an instruction bigger than ~50 bytes 

Closes #17746

### CODE REVIEW

File: libr/asm/p/asm_bf.c [hunk 1]
```diff
@@ -58,80 +58,64 @@ static int disassemble(RAsm *a, RAsmOp *op, const ut8 *buf, int len) {
 	return rep;
 }
 
+static bool _write_asm(RAsmOp *op, int value, int n) {
+	ut8 *opbuf = malloc (n);
```
Comment (wargio) [inline libr/asm/p/asm_bf.c:62]: use calloc and remove the memset at line 66

**Review by wargio — CHANGES_REQUESTED** [review]

File: libr/asm/p/asm_bf.c (hunk 1 above)
Comment (plaets) [reply libr/asm/p/asm_bf.c:62]: What should I use instead of memset here? (Just to be clear - my intention was to fill the buffer with copies of `value`)

Comment (wargio) [reply libr/asm/p/asm_bf.c:62]: oh wait. i'm stupid. sorry. i read that malloc + memset as malloc + bzero behaviour...
Ignore this. i'll merge it as is.

**Review by wargio — APPROVED** [review]

File: libr/asm/p/asm_bf.c (hunk 1 above)
Comment (plaets) [reply libr/asm/p/asm_bf.c:62]: Thanks!

**Pull request merged**